### PR TITLE
Detect when trait is implemented for type and suggest importing it

### LINKED ISSUE #57457: Error suggests implementing private trait instead of importing it.
If I implement a private trait on a public struct, then use a trait method elsewhere, the diagnostic message suggests implementing the already implemented private trait instead of making the trait public and importing it. For example, this code:

```rust
mod m {
    trait T {
        fn f(&self);
    }
    
    pub struct S;
    
    impl T for S {
        fn f(&self) {}
    }
}

fn main() {
    let s = m::S;
    s.f();
}
```

Results in this diagnostic:

```
error[E0599]: no method named `f` found for type `m::S` in the current scope
  --> src/main.rs:15:7
   |
6  |     pub struct S;
   |     ------------- method `f` not found for this
...
15 |     s.f();
   |       ^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `f`, perhaps you need to implement it:
           candidate #1: `m::T`
```

The note should suggest making `m::T` public and `use`ing it at the top level.

### PR DESCRIPTION
**Pull request by estebank**
Fix #57457.

### CODE REVIEW

**Comment by oli-obk** [comment]
Why does this suggestion already work for other errors? Can we somehow simplify something to prevent having to do the same checks 3+ times?

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -2959,26 +2963,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                            }
+                        });
+                    if self
+                        .infcx
+                        .type_implements_trait(trait_info.def_id, args, self.param_env)
+                        .must_apply_modulo_regions()
+                        && param_type.is_none()
+                    {
+                        self.suggest_valid_traits(err, vec![trait_info.def_id], false);
+                    } else {
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs]: done

**Pull request merged**